Add disk IO write mode write-through to advanced settings

summerqB opened the pull request:
and separate the disk IO read mode settings.
This also removes the OS cache settings since it's broken into separate configs now.

summerqB commented:
This basically allows this patch to work https://github.com/arvidn/libtorrent/pull/6703 and sets it by default for all OS.
That patch is supposed to fix https://github.com/arvidn/libtorrent/issues/6561 and some other RC2_0 issues related to disk cache flushing such as corrupted files on macOS(presumably?)

xavier2k6 commented:
>Once applying that patch, **you also need to enable** the option `disk_io_write_mode` = `settings_pack::write_through`.

Main point to note!

Reference: https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1094918506 & https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1114034771

xavier2k6 commented:
very quick test & haven't noted any issues.

xavier2k6 commented:
@Chocobo1 @glassez 
May be a good enough reason to now raise libtorrent requirement to `2.0.6` in `CMakeLists` etc.??

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3489,6 +3523,33 @@ void Session::setDiskQueueSize(const qint64 size)
+#if (LIBTORRENT_VERSION_NUM >= 20006)
+DiskIOWriteMode Session::diskIOWriteMode() const
+{
+  return m_diskIOWriteMode;
```
Wrong indentation here and in method definitions below.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3489,6 +3523,33 @@ void Session::setDiskQueueSize(const qint64 size)
+
+void Session::setDiskIOWriteMode(const DiskIOWriteMode mode)
+{
+  if (mode == m_diskIOWriteMode) return;
```
It's still better to have body of conditional expression in separate line.

glassez commented on src/base/bittorrent/session.h:
```diff
@@ -115,6 +115,23 @@ namespace BitTorrent
+        enum class DiskIOReadMode : int
+        {
+            EnableOSCache = 0,
+            DisableOSCache = 2
```
Why it isn't 1?

glassez commented on src/base/bittorrent/session.h:
```diff
@@ -115,6 +115,23 @@ namespace BitTorrent
+        {
+            EnableOSCache = 0,
+            DisableOSCache = 2,
+            WriteThrough = 3
```
Why not 0, 1, 2?

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -381,7 +381,14 @@ Session::Session(QObject *parent)
     , m_diskCacheTTL(BITTORRENT_SESSION_KEY(u"DiskCacheTTL"_qs), 60)
     , m_diskQueueSize(BITTORRENT_SESSION_KEY(u"DiskQueueSize"_qs), (1024 * 1024))
+#if (LIBTORRENT_VERSION_NUM >= 20006)
+    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"diskIOReadMode"_qs), DiskIOReadMode::EnableOSCache
```
```suggestion
    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"DiskIOReadMode"_qs), DiskIOReadMode::EnableOSCache
```

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -381,7 +381,14 @@ Session::Session(QObject *parent)
+#if (LIBTORRENT_VERSION_NUM >= 20006)
+    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"diskIOReadMode"_qs), DiskIOReadMode::EnableOSCache
+        , clampValue(DiskIOReadMode::EnableOSCache, DiskIOReadMode::DisableOSCache))
+    , m_diskIOWriteMode(BITTORRENT_SESSION_KEY(u"diskIOWriteMode"_qs), DiskIOWriteMode::WriteThrough
```
```suggestion
    , m_diskIOWriteMode(BITTORRENT_SESSION_KEY(u"DiskIOWriteMode"_qs), DiskIOWriteMode::WriteThrough
```

summerqB commented:
> Don't see the added options in advanced section of gui.
>

Should be fixed now.

thalieht commented on src/base/bittorrent/session.h:
```diff
@@ -115,6 +115,23 @@ namespace BitTorrent
+        enum class DiskIOReadMode : int
+        {
+            EnableOSCache = 0,
+            DisableOSCache = 1
```
>Enable = 0
>Disable = 1

Counterintuitive.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -381,7 +381,14 @@ Session::Session(QObject *parent)
+#if (LIBTORRENT_VERSION_NUM >= 20006)
+    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"DiskIOReadMode"_qs), DiskIOReadMode::EnableOSCache
+        , clampValue(DiskIOReadMode::EnableOSCache, DiskIOReadMode::DisableOSCache))
+    , m_diskIOWriteMode(BITTORRENT_SESSION_KEY(u"DiskIOWriteMode"_qs), DiskIOWriteMode::WriteThrough
```
I don't think using `WriteThrough` by default is a good idea.
The code is newly introduced, there might be unknown side-effects/bugs/etc..
I would suggest that the default is switched in later qbt version.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1362,10 +1369,37 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+    switch (diskIOWriteMode())
+    {
+    case DiskIOWriteMode::WriteThrough:
+    default:
```
ditto.

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -115,6 +115,23 @@ namespace BitTorrent
+
+        enum class DiskIOWriteMode : int
+        {
+            EnableOSCache = 0,
```
Same as https://github.com/qbittorrent/qBittorrent/pull/17005/files#r867483426

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -219,8 +224,15 @@ void AdvancedSettings::saveAdvancedSettings()
+    // Disk IO read mode
+    session->setDiskIOReadMode(static_cast<BitTorrent::DiskIOReadMode>(m_comboBoxDiskIOReadMode.currentIndex()));
+    // Disk IO write mode
+    session->setDiskIOWriteMode(static_cast<BitTorrent::DiskIOWriteMode>(m_comboBoxDiskIOWriteMode.currentIndex()));
```
Don't use `currentIndex` for mapping, see how `m_comboBoxOSMemoryPriority` do it properly.

glassez replied:
Currently I strongly believe that the way from #16895 is most reliable.

Chocobo1 reviewed (commented):
Although not strictly required, the webui part is absent. Maybe try adding it?

summerqB commented:
> Although not strictly required, the webui part is absent. Maybe try adding it? (a non-trivial portion of users are using it)

Added!

glassez commented:
@Chocobo1 
How about https://github.com/qbittorrent/qBittorrent/pull/17005#issuecomment-1120395182?

glassez commented on src/webui/api/appcontroller.cpp:
```diff
@@ -315,8 +315,15 @@ void AppController::preferencesAction()
     data[u"disk_cache_ttl"_qs] = session->diskCacheTTL();
     // Disk queue size
     data[u"disk_queue_size"_qs] = session->diskQueueSize();
+#if (LIBTORRENT_VERSION_NUM >= 20006)
```
As discussed earlier, this is unacceptable since it breaks the API.

Chocobo1 commented:
>How about https://github.com/qbittorrent/qBittorrent/pull/17005#issuecomment-1120395182?

I have no opposing reason to say no.

xavier2k6 commented:
> > How about [#17005 (comment)](https://github.com/qbittorrent/qBittorrent/pull/17005#issuecomment-1120395182)?
> 
> I have no opposing reason to say no.

I'll submit a PR....

What about `libtorrent1`, bump to `1.2.16` or leave as is?

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -381,7 +381,14 @@ Session::Session(QObject *parent)
     , m_diskCacheSize(BITTORRENT_SESSION_KEY(u"DiskCacheSize"_qs), -1)
     , m_diskCacheTTL(BITTORRENT_SESSION_KEY(u"DiskCacheTTL"_qs), 60)
     , m_diskQueueSize(BITTORRENT_SESSION_KEY(u"DiskQueueSize"_qs), (1024 * 1024))
+#if (LIBTORRENT_VERSION_NUM >= 20006)
```
There are really two kinds of changes here:
1. Splitting UseOSCache into DiskIOReadMode+DiskIOWriteMode. This should be made in libtorrent version independent way.
2. Adding WriteThrough option to DiskIOWriteMode.

I would make the 2. also in libtorrent version independent way at the qBittorrent level for better compatibility so that WriteThrough option will effectively select DisableOSCache with incompatible libtorrent version.

summerqB replied:
How’ll you deal with users claiming “write through” doesn’t work and seem to be on incompatible version.

glassez replied:
GUI: Don't display "Write through" option at all.

WebUI: Display it as "Write through (libtorrent 2.0.6+)"

When reading settings file or parsing API request just turn it into "Disable OS cache" as I suggested above.

summerqB replied:
Done & done. Please check.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -382,7 +382,10 @@ Session::Session(QObject *parent)
     , m_diskQueueSize(BITTORRENT_SESSION_KEY(u"DiskQueueSize"_qs), (1024 * 1024))
     , m_diskIOType(BITTORRENT_SESSION_KEY(u"DiskIOType"_qs), DiskIOType::Default)
-    , m_useOSCache(BITTORRENT_SESSION_KEY(u"UseOSCache"_qs), true)
+    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"DiskIOReadMode"_qs), DiskIOReadMode::EnableOSCache
+        , clampValue(DiskIOReadMode::DisableOSCache, DiskIOReadMode::EnableOSCache))
```
It seems those values are stored in enum type on disk, so I think there is no need to clamp the values (here and `m_diskIOWriteMode`). The underlying classes will verify the value and reset to default value if it is incompatible/invalid.
```suggestion
    , m_diskIOReadMode(BITTORRENT_SESSION_KEY(u"DiskIOReadMode"_qs), DiskIOReadMode::EnableOSCache)
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -317,8 +317,10 @@ void AppController::preferencesAction()
     data[u"disk_io_type"_qs] = static_cast<int>(session->diskIOType());
-    // Enable OS cache
-    data[u"enable_os_cache"_qs] = session->useOSCache();
+    //Disk IO read mode
```
This also applies to other places/files.
```suggestion
    // Disk IO read mode
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -317,8 +317,10 @@ void AppController::preferencesAction()
-    data[u"enable_os_cache"_qs] = session->useOSCache();
+    //Disk IO read mode
+    data[u"disk_io_read_mode"_qs] = static_cast<int>(session->diskIOReadMode());
+    //Disk IO write mode
```
This also applies to other places/files.
```suggestion
    // Disk IO write mode
```

Chocobo1 commented on src/webui/www/private/views/preferences.html:
```diff
@@ -1052,10 +1052,25 @@
+                    <select id="diskIOWriteMode" style="width: 15em;">
+                        <option value="0">QBT_TR(Disable OS Cache)QBT_TR[CONTEXT=OptionsDialog]</option>
+                        <option value="1">QBT_TR(Enable OS Cache)QBT_TR[CONTEXT=OptionsDialog]</option>
+                        <option value="2">QBT_TR(Write Through(libtorrent >=2.0.6))QBT_TR[CONTEXT=OptionsDialog]</option>
```
```suggestion
                        <option value="2">QBT_TR(Write Through (requires libtorrent >= 2.0.6))QBT_TR[CONTEXT=OptionsDialog]</option>
```

glassez commented on src/gui/advancedsettings.cpp:
```diff
@@ -498,10 +501,21 @@ void AdvancedSettings::loadAdvancedSettings()
+#ifdef QBT_USES_LIBTORRENT2
+    m_comboBoxDiskIOWriteMode.addItem(tr("Write Through"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::WriteThrough));
+#endif
+    m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(session->diskIOWriteMode())));
```
What will be selected in the case of WriteThrough with incompatible libtorrent?

summerqB replied:
Are you talking about the scenario where people have libtorrent <2.0.6? 
I think it will be resolved when we bump the minimum version.

Rootax commented:
(for the record, in my case it doesn't solve https://github.com/arvidn/libtorrent/issues/6561 at all : / )

summerqB commented:
> (for the record, in my case it doesn't solve [arvidn/libtorrent#6561](https://github.com/arvidn/libtorrent/issues/6561) at all : / )

I’d assume you mean that the write through mode set with the build from github actions doesn’t work for you on a windows system?

Can you provide some more info as to you system configuration, working set size set in qBittorrent and the process memory priority?

Rootax commented:
> > (for the record, in my case it doesn't solve [arvidn/libtorrent#6561](https://github.com/arvidn/libtorrent/issues/6561) at all : / )
> 
> I’d assume you mean that the write through mode set with the build from github actions doesn’t work for you on a windows system?
> 
> Can you provide some more info as to you system configuration, working set size set in qBittorrent and the process memory priority?

Hello,

I mean that the performances (constant download speed in my case) doesn't improve. It's still acting like a yoyo (full speed => very slow while it's writing the piece => full speed => very slow while writing, etc). Of course all is ok with lib 1.2.x.

My system is Windows 10 x64, 32gb of ram, i3-8100i cpu, nvme ssd drive (mp510). I try with process memory low, normal, same stuff.  Working size set a 2GB, but the ram usage is always under anyway.  Now I guess this PR is very usefull from some people, but for the people affected by inconsistent speed problem, it's not helping a lot. Still, good job doing this : )

The5kull commented:
I am not experiencing any difference compared to the latest Master-CL build.

The problems I notice are the same as @Rootax mentioned [here](https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1118349404), [here](https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1119325816) and [here](https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1119335136). The HDD is reading a lot of MB/s for an unknown reason which slow the writing to HDD down enormously and eventually the download speeds goes down.

Chocobo1 commented:
@Rootax @The5kull 
Just to be sure we are not missing anything: you have to change the new setting "Disk IO write mode" to "Write Through", otherwise it would be the same as before.

Rootax commented:
> @Rootax @The5kull Just to be sure we are not missing anything: you have to change the new setting "Disk IO write mode" to "Write Through", otherwise it would be the same as before. Also you might be affected by another issue which is recently resolved in PR #16951.

Yes I changed this setting.

For the other issue, from what I read it's impacting mechanical drives, but ssd, not so much. But even with only one hashing thread, same behaviour.

glassez commented:
If [this](https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1119335136) is actually relevant then it is real disadvantage of MMap based I/O subsystem.

The5kull commented:
> > @Rootax @The5kull Just to be sure we are not missing anything: you have to change the new setting "Disk IO write mode" to "Write Through", otherwise it would be the same as before. Also you might be affected by another issue which is recently resolved in PR #16951.
> 
> Yes I changed this setting.
> 
> For the other issue, from what I read it's impacting mechanical drives, but ssd, not so much. But even with only one hashing thread, same behaviour.

@Chocobo1 I had the same experience as @Rootax on 1 hashing thread. Also I did set "Disk IO Write Through", even tried it with POSIX but then qB became frequently less responding ie stuck so that setting went back to default (memory mapped).

> If [this](https://github.com/arvidn/libtorrent/issues/6561#issuecomment-1119335136) is actually relevant then it is real disadvantage of MMap based I/O subsystem.

From what I have read @arvidn is aware of this since it is not windows-only.

glassez commented:
What is current status of this PR?

summerqB commented:
> What is current status of this PR?

Some previous comments suggest that the slow download is due to unnecessary disk reads and this PR doesn't help their situation. So I'm not confident in continuing with this.

Chocobo1 commented:
>So I'm not confident in continuing with this.

The idea of exposing libtorrent settings is still valuable IMO, it gives users a chance to try something out.

summerqB commented:
> > So I'm not confident in continuing with this.
> 
> The idea of exposing libtorrent settings is still valuable IMO, it gives users a chance to try something out.

I think libtorrent should add back RC1_2 style disk I/O and make mmap I/O optional. The former being more predictable and much less buggy. At this point introducing new things in qBit like working set limit and PRs like this one which are half baked solutions, might end up with unexpected pitfalls.

Rootax commented:
> > > So I'm not confident in continuing with this.
> > 
> > 
> > The idea of exposing libtorrent settings is still valuable IMO, it gives users a chance to try something out.
> 
> I think libtorrent should add back RC1_2 style disk I/O and make mmap I/O optional. The former being more predictable and much less buggy. At this point introducing new things in qBit like working set limit and PRs like this one which are half baked solutions, might end up with unexpected pitfalls.


I don't think it will happen, the main "improvement" was 
supposed to be the new i/o system, no more cache to manage,etc...

glassez commented:
> I think libtorrent should add back RC1_2 style disk I/O and make mmap I/O optional. The former being more predictable and much less buggy.

This is a huge drawback of libtorrent 2.0, which apparently will not allow it to replace libtorrent 1.2, even despite the support of BitTorrent v2, which no one needs at the cost of such an unsuccessful I/O subsystem. After all, memory problems are far from its only drawback. This can include problems with network drives, problems with the latest macOS, etc.

glassez commented:
> I don't think it will happen, the main "improvement" was
> supposed to be the new i/o system, no more cache to manage,etc...

Yes, but it seems to have failed.

arvidn commented:
> > I don't think it will happen, the main "improvement" was
> > supposed to be the new i/o system, no more cache to manage,etc...
> 
> Yes, but it seems to have failed.

Yes, at least for some people in some scenarios. The new disk I/O subsystem does not preclude using a threaded `pwritev()` and `preadv()`  subsystem. It just hasn't been written yet.

summerqB commented:
> I don't think it will happen, the main "improvement" was supposed to be the new i/o system, no more cache to manage,etc...
>

I'm not asking for removal of the "improvements" but adding back of removed system in a way so it can be configured on the fly as it was better and had less issues compared to now.

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -115,6 +115,21 @@ namespace BitTorrent
         };
         Q_ENUM_NS(BTProtocol)
 
+        enum class DiskIOReadMode : int
```
Wrong place, please sort the enums in alphabetical order.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -508,10 +511,19 @@ void AdvancedSettings::loadAdvancedSettings()
+    // Disk IO write mode
+    m_comboBoxDiskIOWriteMode.addItem(tr("Disable OS Cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::DisableOSCache));
+    m_comboBoxDiskIOWriteMode.addItem(tr("Enable OS Cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::EnableOSCache));
+    m_comboBoxDiskIOWriteMode.addItem(tr("Write Through (requires libtorrent >= 2.0.6)"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::WriteThrough));
```
IMO (for GUI only) it isn't good to use a such wording, better use `#if` to enable/disable it at compile time. Otherwise linux user might complain that they upgraded the libtorrent version from 2.0.5 to 2.0.6 but the option is not working.

summerqB replied:
then what'll be selected incase someone downgrades?
https://github.com/qbittorrent/qBittorrent/pull/17005#discussion_r873131347

summerqB replied:
I changed the wording a bit. See if that helps.

glassez replied:
> Otherwise linux user might complain that they upgraded the libtorrent version from 2.0.5 to 2.0.6 but the option is not working.

Why not increase requirements up to libtorrent 2.0.6 (for 2.0 branch)?
>better use #if to enable/disable it at compile time. 

IMO, it's preferred way for GUI.
@summerqB, don't you still understand why we can't do the same in WebUI?


> then what'll be selected incase someone downgrades?
> [#17005 (comment)](https://github.com/qbittorrent/qBittorrent/pull/17005#discussion_r873131347)

The mentioned comment is about differences between the same qBittorrent built with different branches of libtorrent.
I mean if option in config is write-through but libtorrent is 1.2 (e.g. libtorrent 2.0 based build was used previously) you should explicitly replace it with one of the supported options.

Chocobo1 replied:
>then what'll be selected incase someone downgrades?

From previous discussion with other developers, we reached an conclusion: It is often near impossible to provide reliable behavior for downgrades so we usually won't talk about it. We would only consider upgrade scenario.

As for this specific code: qbt will detect incompatible setting and use the default value (which is `EnableOSCache`).

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1619,10 +1620,34 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+    default:
+        settingsPack.set_int(lt::settings_pack::disk_io_write_mode, lt::settings_pack::enable_os_cache);
+        break;
+    case DiskIOWriteMode::WriteThrough:
+#ifdef QBT_USES_LIBTORRENT2
+        settingsPack.set_int(lt::settings_pack::disk_io_write_mode, lt::settings_pack::write_through);
+#else
+        settingsPack.set_int(lt::settings_pack::disk_io_write_mode, lt::settings_pack::disable_os_cache);
+#endif
+        break;
```
```suggestion
#ifdef QBT_USES_LIBTORRENT2
    case DiskIOWriteMode::WriteThrough:
        settingsPack.set_int(lt::settings_pack::disk_io_write_mode, lt::settings_pack::write_through);
        break;
#endif
```

summerqB replied:
what will happen when someone selects writethrough from WebUI or when they downgrade.

Chocobo1 replied:
>what will happen when someone selects writethrough from WebUI or when they downgrade.

Same as https://github.com/qbittorrent/qBittorrent/pull/17005#discussion_r916769794. Additionally it will fall into `switch()` `default:` case.

summerqB replied:
Done.

summerqB commented:
I think I've addressed all concerns.
People using libtorrent 1.2.x won't see the option in GUI. If someone set it to write through when using 2.0.x, after downgrading that option will change to Disable OS cache automatically.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -508,10 +511,26 @@ void AdvancedSettings::loadAdvancedSettings()
+    // Disk IO write mode
+    m_comboBoxDiskIOWriteMode.addItem(tr("Disable OS cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::DisableOSCache));
+    m_comboBoxDiskIOWriteMode.addItem(tr("Enable OS cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::EnableOSCache));
+#ifdef QBT_USES_LIBTORRENT2
+    m_comboBoxDiskIOWriteMode.addItem(tr("Write-through"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::WriteThrough));
+    m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(session->diskIOWriteMode())));
+#else
+    if (session->diskIOWriteMode() == BitTorrent::DiskIOWriteMode::WriteThrough)
+        m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(BitTorrent::DiskIOWriteMode::DisableOSCache)));
+    else
+        m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(session->diskIOWriteMode())));
+#endif
```
I meant like this:
```suggestion
#ifdef QBT_USES_LIBTORRENT2
    m_comboBoxDiskIOWriteMode.addItem(tr("Write-through"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::WriteThrough));
#endif
    m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(session->diskIOWriteMode())));
```

summerqB replied:
I did this initially but then glassez commented about downgrading issues.

summerqB replied:
Done.

summerqB commented:
I've updated the PR again. Now when user downgrades libtorrent, nothing is selected. So it most likely falls back to default case.

![image](https://user-images.githubusercontent.com/79678786/178649539-20eb91c3-ea2c-4d34-9288-66cbf1d9c3b2.png)

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1619,10 +1620,33 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+        settingsPack.set_int(lt::settings_pack::disk_io_write_mode, lt::settings_pack::write_through);
+        break;
+#endif
+
```
Please remove this empty line.

Chocobo1 reviewed (commented):
One last comment and then it is ready

Chocobo1 commented:
Appveyor is failing, so either its libtorrent needs to be updated or qbt needs to bump its supported libtorrent version number.

glassez commented:
> Now when user downgrades libtorrent, nothing is selected. So it most likely falls back to default case.

And what will be saved to the settings file if the user applies some changes?

I would still prefer to see the option that is effectively used.

It seems that we are confused about what we want to get in the end.
The cornerstone of the problem is the Web API, which should be stable within its version, regardless of the libtorrent version used, etc., i.e. it should have a single set of methods/endpoints and supported parameters.
But at the application core level, support for certain functions may differ. So, for example, Write-through is not supported by the libtorrent < 2.0.6 based builds, so it is an invalid value. If it is contained in the configuration file or in the Web API request, then it should be treated as any other invalid value (replaced by the default value). The WebAPI can simply return the runtime error code in this case (for example, HTTP Conflict status). As for the UI, the GUI simply should not display invalid options. The WebUI is more difficult. It can either display all options, but have warnings that some of them are not supported by some version of libtorrent, or you can try to change it so that it behaves like GUI, i.e. it receives the current version of libtorrent and displays only those options that are available.

Chocobo1 commented:
>As for the UI, the GUI simply should not display invalid options. 

It is missing a conditional at the enum definition which will obstruct the invalid value detection, and I'll handle it.

>If it is contained in the configuration file or in the Web API request, then it should be treated as any other invalid value (replaced by the default value). 

It seems it already do that.

>The WebAPI can simply return the runtime error code in this case (for example, HTTP Conflict status). 

It is cumbersome to do that in the current code, so IMO let just let it silently fall back to default value for now.

Other comments should be already handled correctly.
This PR is near to finish, I plan to took over and handle it myself.
